mithril-stm/src/stm.rs

```diff
@@ -350,38 +384,64 @@ impl StmInitializer {
     }
 }
 
-impl<D: Clone + Digest + FixedOutput> StmSigner<D> {
-    /// This function produces a signature following the description of Section 2.4.
+impl StmSigner {
     /// Once the signature is produced, this function checks whether any index in `[0,..,self.params.m]`
     /// wins the lottery by evaluating the dense mapping.
     /// It records all the winning indexes in `Self.indexes`.
-    /// If it wins at least one lottery, it stores the signer's merkle tree index. The proof of membership
-    /// will be handled by the aggregator.
-    pub fn sign(&self, msg: &[u8]) -> Option<StmSig> {
-        let msgp = self
-            .closed_reg
-            .merkle_tree
-            .to_commitment_batch_compat()
-            .concat_with_msg(msg);
-        let sigma = self.sk.sign(&msgp);
+    pub fn sign(&self, msg: &[u8], total_stake: Stake) -> Option<StmSig> {
+        let sigma = self.sk.sign(msg);
+
+        let indexes = self.check_lottery(msg, &sigma, total_stake);
+
+        if !indexes.is_empty() {
+            Some(StmSig {
+                sigma,
+                indexes,
+                signer_index: self.mt_index,
+            })
+        } else {
+            None
+        }
+    }
 
-        // Check which lotteries are won
+    /// Checks whether the indices won the lottery.
+    /// Collects and returns the winning indices.
+    pub fn check_lottery(&self, msg: &[u8], sigma: &Signature, total_stake: Stake) -> Vec<u64> {
         let mut indexes = Vec::new();
         for index in 0..self.params.m {
             if ev_lt_phi(
                 self.params.phi_f,
-                sigma.eval(&msgp, index),
+                sigma.eval(msg, index),
                 self.stake,
-                self.closed_reg.total_stake,
+                total_stake,
             ) {
                 indexes.push(index);
             }
         }
-        if !indexes.is_empty() {
+        indexes
+    }
+}
+
+impl<D: Clone + Digest + FixedOutput> StmSignerAvk<D> {
+    /// This function produces a signature following the description of Section 2.4.
+    /// Once the signature is produced, this function checks whether any index in `[0,..,self.params.m]`
+    /// wins the lottery by evaluating the dense mapping.
+    /// It records all the winning indexes in `Self.indexes`.
+    /// If it wins at least one lottery, it stores the signer's merkle tree index. The proof of membership
+    /// will be handled by the aggregator.
+    pub fn sign(&self, msg: &[u8]) -> Option<StmSig> {
+        let msgp = self
+            .closed_reg
+            .merkle_tree
+            .to_commitment_batch_compat()
+            .concat_with_msg(msg);
+        let signature = StmSigner::sign(&self.stm_signer, &msgp, self.closed_reg.total_stake)?;
+
+        if !signature.indexes.is_empty() {
```

Comment:
The check if `indexes.is_empty` is already performed inside the `sign` function. With the `?` you are implicitly saying that if there are no indices, return `None`, and if there are, continue.